### PR TITLE
hyprstream-decision-backbone: System One P1.2 backbone conversion (LLM2Vec, softmax layers only) + ModernBERT baseline arm

### PR DESCRIPTION
**Pull request by ewindisch**
System One program, DAG node **P1.2** (Wave 1, hard-dep on P1.1 #1659 — merged).

## Scope → acceptance mapping (plan v1.7.2, P1.2 section)

| Plan requirement | Where |
|---|---|
| LLM2Vec-style bidirectional conversion, **softmax layers only** | `Qwen35Encoder` + `MaskPolicy::Bidirectional` (`crates/hyprstream-decision-backbone/src/qwen35.rs`, `backbone.rs`). Test: `softmax_layers_attend_bidirectionally_after_conversion` (future tokens change earlier states; causal policy control does not). |
| 18 GDN layers stay causal (partial precedent) | `gdn.rs`: torch-native chunked gated delta rule, causal by construction, autograd-safe (out-of-place within-chunk correction). Test: `gdn_layers_stay_causal_under_the_bidirectional_conversion` (prefix states bit-invariant to future-token perturbation). |
| Amputate LM head | No `lm_head`/`embed_out`/`mtp` parameters are ever allocated; `Qwen35Encoder::amputated_lm_head_names()` lists the prefixes to drop at checkpoint conversion. |
| Attach the shared scorer (P1.1 integration) | `DecisionModel<B: Backbone>` (`model.rs`): backbone + `SharedScorer` in one `VarStore`, one forward pass → typed per-question distributions. Tests: end-to-end distributions on both arms, joint checkpoint round-trip, KL distillation loss backprops into backbone **and** scorer. |
| Review nit: `SharedScorer::new` takes `nn::Path` | Fixed in `hyprstream-decision-head/src/scorer.rs` (was `&VarStore`/`vs.root()`; caller now scopes `vs.root().sub("scorer")` — no backbone param collision; test `scorer_params_do_not_collide_with_backbone_params`). |
| ModernBERT-large baseline arm | `modernbert.rs`: encoder-only ModernBERT (alternating local/global attention, GeGLU, biased LayerNorms, dual-theta RoPE) behind the same `Backbone` trait; `ModernBertConfig::large()` is the S3 reference shape. Test: local-window masking at the attention sublayer. |
| Torch-native GDN path pinned; fla behind Triton ≥3.7 gate | `gdn.rs` is pure tensor ops (zero Triton/fla); crate docs record the fallback ranking (native → fla@Triton≥3.7 → MI300X rental → 5090-only). |
| Deliverable: dual-backend pin (CUDA/ROCm) | `backend.rs` + crate docs: one source tree, CUDA build → RTX 5090 dev, ROCm build → MI210 volume; both surface via `Device::Cuda`; selection is link-time via `LIBTORCH`. |
| Deliverable: training-framework/DDP decision | Crate docs: single-GPU tch-rs first (no RCCL in hyprstream today — S6c inventory flag); libtorch-distributed data-parallel within one 4×xGMI group at volume; cross-group embarrassingly parallel; Python-framework and FSDP/TP options rejected with reasons. |
| Deliverable: 4-step smoke test | `src/bin/hyprstream-backbone-smoke.rs`: (1) rocm-smi/nvidia-smi probe → (2) libtorch FFI probe → (3) one fwd/bwd optimizer step at 2k ctx with MFU sizing reference → (4) fla leg reported gated-skipped. **Run on this host: PASS — RTX 5090 detected, FFI functional, 491 ms / 4168 tok/s at 2048 ctx (tiny config) on CUDA.** |
| Hybrid fallback pre-priced | Crate docs: generalist scorer ships as-is; per-family calibration anchoring is pure P2.2 post-hoc fitting (params persisted as P0.3 rows) — promoting the fallback is a configuration event, no re-architecture. |

## Boundaries

- New library crate: **Apache-2.0**, registered in `.github/license-boundary.toml` (apache_packages + permissive_roots); `check_license_boundary.py` passes. The GDN implementation is an independent encoder-side implementation of the public chunked-delta-rule algorithm — no AGPL `hyprstream-inference` dependency (license gate verified).
- Workspace member added; canonical hakari sequence run (ungate → `hakari generate` → `manage-deps` → re-gate) — byte-clean ("no changes detected").
- No `127.0.0.1` literals; no AI-attribution trailers.

## Validation

- `cargo test -p hyprstream-decision-backbone -p hyprstream-decision-head`: 13 + 13 pass.
- Full-workspace `cargo clippy --workspace --all-targets -- -D warnings`: clean (also release clippy via pre-commit hook).
- Smoke binary run on RTX 5090: PASS (output in PR discussion).
- ModernBERT checkpoint parity against the reference implementation is explicitly a P1.4-time task (needs real weights); the arm here is the encoder-relevant subset, shape- and training-verified.

### CODE REVIEW

File: crates/hyprstream-decision-backbone/src/modernbert.rs [hunk 1]
```diff
@@ -0,0 +1,571 @@
...
+        let qkv = flat
+            .apply(&self.wqkv)
+            .view([batch, seq, 3, self.num_heads, self.head_dim])
+            .permute([2, 0, 3, 1, 4]); // [3, B, H, T, d]
+        let q = rope.apply(&qkv.get(0));
+        let k = rope.apply(&qkv.get(1));
+        let v = qkv.get(2);
+        let scale = 1.0 / (self.head_dim as f64).sqrt();
+        let scores = q.matmul(&k.transpose(-1, -2)).to_kind(Kind::Float) * scale
+            + additive_mask.to_kind(Kind::Float);
```
[the remark this answers is not in the document]
Comment (ewindisch) [reply crates/hyprstream-decision-backbone/src/modernbert.rs]: Already fixed at this PR's head in d4be51297 ("chunked attention (P1 memory) + checkpoint prefix remap (P2)", pushed after this thread was filed): local layers compute attention over band-sliced key windows per query chunk instead of materializing the dense [B, 16, seq, seq] score tensor, restoring sliding-window memory complexity for the ModernBERT arm. Independently verified in review round 3 (system-one/reviews/pr1666-r3.md): math derivation of band coverage, a load-bearing equivalence test, and a mutation check that fails when the band is shrunk by one key. No further change needed; resolving.

File: crates/hyprstream-decision-backbone/src/qwen35.rs [hunk 2]
```diff
@@ -0,0 +1,630 @@
...
+            hidden_dim: cfg.hidden_size,
+        })
+    }
+
+    /// The parameter name prefixes a source checkpoint carries that this encoder
+    /// deliberately does not allocate — the LM head amputation. Drop every tensor
+    /// under these prefixes when converting a generative checkpoint into the
+    /// decision model's `VarStore` (extra tensors would otherwise fail the load).
+    pub fn amputated_lm_head_names() -> &'static [&'static str] {
+        &["lm_head.", "embed_out.", "mtp."]
```
[the remark this answers is not in the document]
Comment (ewindisch) [reply crates/hyprstream-decision-backbone/src/qwen35.rs:469]: Already fixed at this PR's head in d4be51297: Qwen35Encoder::remap_checkpoint_name now handles nested checkpoint layouts, matching the existing qwen3_5.rs:1517-1537 remapping of model.language_model.* and model.visual.* prefixes so Qwen3.5 Instruct checkpoints load into the encoder. Independently verified in review round 3 (pr1666-r3.md) against the live HF index. No further change needed; resolving.
